plugins/pulp_rpm/plugins/importers/yum/pulp_solv.py

```diff
@@ -321,299 +264,285 @@ def rpm_dependency_conversion(solvable, unit_name, unit_flags, unit_evr,
+          dep_or_rel(solv.Dep, solve.REL_WITH, stream_dep(name, stream))
+          dep_or_rel(solv.Dep, solve.REL_WITHOUT, stream_dep(name, stream))
+        """
+        dep.Rel(op, rel) if dep is not None else rel
+
+    # Check each dependency dict in dependency_list and generate the solv requirements.
+    reqs = None
+    for dep_dict in dependency_list:
+        require = None
+        for name, streams in dep_dict.items():
+            if name == 'platform':
+                # no need to fake the platform (streams) later on
+                continue
+            name = name.encode('utf8')
+
+            # The req_pos will store solv.Dep expression for "positive" requirements.
+            # That is the case of 'gtk': ['1', '2'].
+            # The req_neg will store negative requirements like 'gtk': ['-1', '-2'].
+            req_pos = req_neg = None
+
+            # For each stream in `streams` for this dependency, generate the
+            # module(name:stream) solv.Dep and add REL_OR relations between them.
+            for stream in streams:
+                stream = stream.encode('utf8')
+
+                if stream.startswith("-"):
+                    req_neg = dep_or_rel(req_neg, solv.REL_OR, stream_dep(name, stream[1:]))
+                else:
+                    req_pos = dep_or_rel(req_pos, solv.REL_OR, stream_dep(name, stream))
+
+            # Generate the module(name) solv.Dep.
+            req = pool.Dep("module({})".format(name))
+
+            # Use the REL_WITH for positive requirements and REL_WITHOUT for negative
+            # requirements.
+            if req_pos is not None:
+                req = req.Rel(solv.REL_WITH, req_pos)
+            elif req_neg is not None:
+                req = req.Rel(solv.REL_WITHOUT, req_neg)
+
+            # And in the end use AND between the last name:[streams] and the current one.
+            require = dep_or_rel(require, solv.REL_AND, req)
+
+        # NOTE: In the original algorithm, this was an OR operation. We don't want only one
+        # set of deps (for one platform), we want the deps for all platforms. Hence, solv.REL_AND.
+        reqs = dep_or_rel(reqs, solv.REL_AND, require)
+
+    module_solvable.add_deparray(solv.SOLVABLE_REQUIRES, reqs)
+
+
+def module_defaults_unit_to_solvable(solv_repo, unit):
+    """Convert a Pulp Module Default dict to a libsolv solvable.
+
+    :param solv_repo: the repository the unit is being added into
+    :type solv_repo: solv.Repo
+
+    :param unit: the unit being converted
+    :type unit: pulp_rpm.plugins.models.Model
+
+    :return: the solvable created.
+    :rtype: solv.Solvable
     """
-    pool = solvable.repo.pool
+    solvable = solv_repo.add_solvable()
+    solvable.evr = ''
+    solvable.arch = ''
+
+    name = unit.get('name', None)
+    if name is not None:
+        name = name.encode('utf-8')
+
+    stream = unit.get('stream', None)
+    if stream is not None:
+        stream = stream.encode('utf-8')
+
     if stream:
```

Comment:
I think one `if` statement above is enough and covers this case as well.

Reply:
I tried to avoid mixing the type safety guards with the logic deliberately to make them as obvious as possible so I'd rather leave this as-is.  Plus, a unicode-encoded empty string would be able to slip through a combined guard and I'm not sure if that would cause problems.

Reply:
(I did violate that in a couple of places like ```module_solvable_name``` where it would have otherwise been really verbose and it was obvious anyways what was going on.